### ISSUE TITLE
How to do fine-tuning?

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/candle#891]
Hello everybody,

I was looking through the Candle examples and cannot seem to find an example of fine-tuning for Llama. It appears the only example present is for training from scratch. How should I fine-tune a pretrained model on my own data? Or, more generally, how should I fine tune a model that it loaded from a safetensor file (and whose VarBuilder is immutable as discussed in #883)?

Thanks!
Eric

**Comment by LaurentMazare** [comment]
If you look at the mnist-training example, you can see that we could load an existing checkpoint in a `VarMap` and retrain from this point. See [here](https://github.com/huggingface/candle/blob/82a98f6da0e0d8273bcb6cf5766bc25a58843f80/candle-examples/examples/mnist-training/main.rs#L127) for loading the data (and somewhere else in the file for writing it down).
If you only want to train for a subset of variables, you can probably just `detach` them or use a custom backend - if this use code is fairly common we could consider adding a specific backend but that would probably be tied to a single file format so better if we can avoid and stay generic.

**Comment by EricLBuehler** [comment]
Thank you! For future reference, creating a VarBuilder from a map and then calling .load to load a file will allow for fine-tuning.

**Comment by LaurentMazare** [comment]
Yes, we should certainly get better at documenting these. E.g. the mnist-training example should have some form of tutorial version where we would explain the different features and how it's supposed to work as it's super hard to guess.

**Comment by mohamed-180** [comment]
Hello everybody,

Is there away to freeze all Var Tensors in the VarMap like the below snippet ?
means something like implement the `Iterator` trait and detach the contained tensors from the graph and add a Var which can be trained !!!

```
# Freeze all the pre-trained layers
for param in model.parameters():
    param.requires_grad = False

```
